In [1]:
import numpy as np

from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    classification_report,
    roc_auc_score
)

# Create dataset
X, y = make_classification(
    n_samples=1000,
    n_features=10,
    n_classes=2,
    weights=[0.7, 0.3],
    random_state=42
)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

# Train model
model = LogisticRegression()

model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

# Probability predictions
y_prob = model.predict_proba(X_test)[:, 1]

In [2]:
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
auc = roc_auc_score(y_test, y_prob)

print("Accuracy :", accuracy)
print("Precision:", precision)
print("Recall   :", recall)
print("F1 Score :", f1)
print("ROC-AUC  :", auc)

Accuracy : 0.895
Precision: 0.8823529411764706
Recall   : 0.75
F1 Score : 0.8108108108108109
ROC-AUC  : 0.9477380952380953


In [3]:
cm = confusion_matrix(y_test, y_pred)

print(cm)

[[134   6]
 [ 15  45]]


In [4]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.90      0.96      0.93       140
           1       0.88      0.75      0.81        60

    accuracy                           0.90       200
   macro avg       0.89      0.85      0.87       200
weighted avg       0.89      0.90      0.89       200



In [5]:
y_pred = model.predict(X_test)

In [6]:
threshold = 0.8

y_pred_strict = (y_prob >= threshold).astype(int)

print("Precision:", precision_score(y_test, y_pred_strict))
print("Recall   :", recall_score(y_test, y_pred_strict))

Precision: 0.967741935483871
Recall   : 0.5


In [7]:
threshold = 0.2

y_pred_sensitive = (y_prob >= threshold).astype(int)

print("Precision:", precision_score(y_test, y_pred_sensitive))
print("Recall   :", recall_score(y_test, y_pred_sensitive))

Precision: 0.7402597402597403
Recall   : 0.95


In [8]:
tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()

specificity = tn / (tn + fp)

print("Specificity:", specificity)

Specificity: 0.9571428571428572


In [9]:
thresholds = [0.2, 0.5, 0.8]

for threshold in thresholds:

    predictions = (y_prob >= threshold).astype(int)

    print(
        f"Threshold={threshold} | "
        f"Precision={precision_score(y_test, predictions):.3f} | "
        f"Recall={recall_score(y_test, predictions):.3f} | "
        f"F1={f1_score(y_test, predictions):.3f}"
    )

Threshold=0.2 | Precision=0.740 | Recall=0.950 | F1=0.832
Threshold=0.5 | Precision=0.882 | Recall=0.750 | F1=0.811
Threshold=0.8 | Precision=0.968 | Recall=0.500 | F1=0.659


In [10]:
import numpy as np

from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

# Create regression dataset
X, y = make_regression(
    n_samples=200,
    n_features=3,
    noise=15,
    random_state=42
)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# Create model
model = LinearRegression()

# Train
model.fit(X_train, y_train)

# Predictions
y_pred = model.predict(X_test)

In [11]:
mae = mean_absolute_error(y_test, y_pred)

mse = mean_squared_error(y_test, y_pred)

rmse = np.sqrt(mse)

r2 = r2_score(y_test, y_pred)

print("MAE :", mae)
print("MSE :", mse)
print("RMSE:", rmse)
print("R²  :", r2)

MAE : 13.864257365888227
MSE : 298.4033866292122
RMSE: 17.274356330387892
R²  : 0.9667410895959422


In [12]:
errors = y_test - y_pred

print("First 10 Actual Values:")
print(y_test[:10])

print("\nFirst 10 Predictions:")
print(y_pred[:10])

print("\nFirst 10 Errors:")
print(errors[:10])

First 10 Actual Values:
[ 165.73353135   85.36700115  -13.09128267  -86.12418336   69.98334659
 -149.95224496  -74.25750479  -17.58757696   33.34470228  -18.57792363]

First 10 Predictions:
[ 167.66228909   63.53807548  -52.55780848 -126.13202354   86.25982098
 -141.87201695  -66.12433473  -26.25409059   56.34775398  -27.99241352]

First 10 Errors:
[ -1.92875775  21.82892568  39.46652581  40.00784018 -16.27647439
  -8.08022801  -8.13317006   8.66651363 -23.0030517    9.4144899 ]


# ML Evaluation & Selection — 01. Evaluation Metrics

## 1. Model Evaluation

Model evaluation measures how well a machine learning model performs on **unseen data**.

The main goal is to check **generalization**.

```text
Training Data
      ↓
Train Model
      ↓
Unseen Data
      ↓
Evaluate Performance
```

### Training vs Test Performance

```text
High Training Score + Low Test Score
→ Overfitting

Low Training Score + Low Test Score
→ Underfitting

Good Training Score + Good Test Score
→ Better Generalization
```

---

# 2. Classification Metrics

Classification metrics are used when the target is a **class/category**.

Examples:

```text
Spam / Not Spam
Fraud / Not Fraud
Disease / No Disease
Cat / Dog
```

The basic confusion matrix is:

```text
                 Actual
              Negative  Positive

Predicted
Negative        TN        FN

Positive        FP        TP
```

* **TP (True Positive)** → Model predicted Positive and it was Positive
* **TN (True Negative)** → Model predicted Negative and it was Negative
* **FP (False Positive)** → Model predicted Positive but it was Negative
* **FN (False Negative)** → Model predicted Negative but it was Positive

---

# 3. Accuracy

Accuracy measures the proportion of **all predictions that are correct**.

$$
Accuracy = \frac{TP + TN}{TP + TN + FP + FN}
$$

### Example

```text
100 predictions
95 predictions are correct
```

$$
Accuracy = \frac{95}{100} = 0.95 = 95\%
$$

### When to use Accuracy

Accuracy is useful when:

* Classes are reasonably balanced
* FP and FN have similar importance

Example:

```text
500 Cat
500 Dog
```

Accuracy is meaningful because neither class dominates the dataset.

### Problem with Accuracy

Accuracy can be misleading with **imbalanced data**.

Example:

```text
990 → Normal
10  → Fraud
```

A model that predicts everything as Normal gets:

```text
990 / 1000 = 99% Accuracy
```

But:

```text
Fraud detected = 0 / 10
```

So the model has **99% accuracy but completely fails at detecting fraud**.

---

# 4. Precision

Precision answers:

> **Of everything the model predicted as Positive, how many were actually Positive?**

$$
Precision = \frac{TP}{TP + FP}
$$

Think:

```text
Model predicted Positive
          ↓
How many were actually Positive?
          ↓
       Precision
```

### Example

The model predicts 100 transactions as fraud.

```text
80 → Actually Fraud
20 → Actually Normal
```

$$
Precision = \frac{80}{80 + 20}
$$

$$
Precision = 0.80 = 80\%
$$

So **80% of the model's Positive predictions were correct**.

### When to use Precision

Use Precision when **False Positives are costly**.

```text
FP
↓
Model says Positive
but actually Negative
```

### Example — Spam Detection

```text
Normal Email
     ↓
Model says Spam
     ↓
False Positive
```

Too many False Positives can cause important emails to be incorrectly classified as spam.

### Example — Fraud Detection

If incorrectly blocking a legitimate customer's transaction is expensive, we may want higher Precision.

### Remember

> **Precision → Can I trust my Positive prediction?**

---

# 5. Recall

Recall answers:

> **Of all the samples that were actually Positive, how many did the model find?**

$$
Recall = \frac{TP}{TP + FN}
$$

Think:

```text
Actually Positive
       ↓
How many did the model find?
       ↓
      Recall
```

### Example

There are 100 actual fraud cases.

```text
90 → Detected
10 → Missed
```

$$
Recall = \frac{90}{90 + 10}
$$

$$
Recall = 0.90 = 90\%
$$

The model found **90% of all actual fraud cases**.

### When to use Recall

Use Recall when **False Negatives are costly**.

```text
FN
↓
Actually Positive
but model says Negative
```

### Example — Disease Detection

```text
100 patients actually have disease

95 → Detected
5  → Missed
```

$$
Recall = \frac{95}{95 + 5} = 95\%
$$

Missing a sick patient can be much more costly than incorrectly flagging a healthy patient.

Therefore, **Recall is often very important in disease screening**.

### Remember

> **Recall → Did I find the Positives?**

---

# 6. Precision vs Recall

| Metric        | Main Question                          | Main Concern    |
| ------------- | -------------------------------------- | --------------- |
| **Precision** | When I predict Positive, am I correct? | False Positives |
| **Recall**    | Did I find the actual Positives?       | False Negatives |

### If False Positives are costly

```text
FP is costly
     ↓
Focus on Precision
```

### If False Negatives are costly

```text
FN is costly
     ↓
Focus on Recall
```

### If both matter

```text
Precision + Recall
        ↓
    F1 Score
```

---

# 7. Precision–Recall Trade-off

Most classification models produce a **probability**.

Example:

```text
Fraud Probability = 0.82
```

We choose a threshold:

```text
Probability >= 0.5
→ Positive
```

Changing the threshold changes the model's predictions.

### Higher Threshold

```text
Threshold ↑
     ↓
Fewer Positive predictions
     ↓
Precision tends to ↑
Recall tends to ↓
```

### Lower Threshold

```text
Threshold ↓
     ↓
More Positive predictions
     ↓
Recall tends to ↑
Precision tends to ↓
```

Therefore, the classification threshold can be adjusted according to the problem.

---

# 8. F1 Score

F1 Score combines **Precision and Recall** into a single metric.

$$
F1 = 2 \times \frac{Precision \times Recall}{Precision + Recall}
$$

F1 Score is the **harmonic mean** of Precision and Recall.

### Example

```text
Precision = 90%
Recall    = 90%
```

Both are high, so F1 will also be high.

But:

```text
Precision = 99%
Recall    = 20%
```

F1 will be much lower because Recall is poor.

### When to use F1

F1 is useful when:

* Classes are imbalanced
* Both FP and FN matter
* We need a balance between Precision and Recall

### Remember

> **F1 → Balance between Precision and Recall**

---

# 9. Specificity

Specificity measures how well the model identifies **actual Negative cases**.

$$
Specificity = \frac{TN}{TN + FP}
$$

Think:

```text
Actual Negative
      ↓
How many did the model correctly identify?
      ↓
  Specificity
```

### Remember

```text
Recall
→ Finding actual Positives

Specificity
→ Correctly identifying actual Negatives
```

---

# 10. ROC-AUC

ROC-AUC measures how well a model can **separate Positive and Negative classes across different classification thresholds**.

It is not simply the percentage of correct predictions.

```text
AUC ≈ 1.0
→ Excellent class separation

AUC ≈ 0.5
→ Random-like

AUC < 0.5
→ Worse than random
```

### Important

ROC-AUC evaluates the model's ability to **distinguish between classes across thresholds**.

---

# 11. Regression Metrics

Regression models predict **continuous numerical values**.

Example:

```text
Actual     Predicted

100        95
200        210
300        280
```

The difference between actual and predicted values is called the **error/residual**.

```text
Error = Actual - Predicted
```

---

# 12. MAE — Mean Absolute Error

MAE measures the **average absolute error**.

$$
MAE = \frac{\sum |y - \hat{y}|}{n}
$$

Where:

```text
y     → Actual value
ŷ     → Predicted value
n     → Number of samples
```

### Example

```text
Errors:

5
10
20
```

$$
MAE = \frac{5 + 10 + 20}{3}
$$

$$
MAE = 11.67
$$

### Important

MAE is expressed in the **same units as the target**.

Lower MAE is better.

### Use MAE when

You want a straightforward measure of the **average prediction error**.

---

# 13. MSE — Mean Squared Error

MSE calculates the average of **squared errors**.

$$
MSE = \frac{\sum (y - \hat{y})^2}{n}
$$

Squaring the errors makes **large errors much more important**.

Example:

```text
Error = 2
→ 2² = 4

Error = 10
→ 10² = 100
```

So a large error has a much larger effect on MSE.

Lower MSE is better.

---

# 14. RMSE — Root Mean Squared Error

RMSE is the square root of MSE.

$$
RMSE = \sqrt{MSE}
$$

RMSE:

* Penalizes large errors
* Is expressed in the same units as the target
* Is easier to interpret than MSE

Lower RMSE is better.

### Quick Comparison

```text
MAE
→ Average absolute error

MSE
→ Strongly penalizes large errors

RMSE
→ Same idea as MSE + returns to original units
```

---

# 15. R² — R-Squared

R² measures how much of the variation in the target is explained by the model compared with a **mean-based baseline**.

General interpretation:

```text
R² = 1
→ Perfect prediction

R² = 0
→ No improvement over mean baseline

R² < 0
→ Worse than mean baseline
```

Higher R² is generally better.

---

# 16. Choosing the Right Metric

There is **no single best metric for every ML problem**.

Ask:

> **What type of mistake is more costly?**

```text
                 ML Problem
                     ↓
             What mistake matters?
                ↙           ↘
              FP             FN
              ↓              ↓
          Precision        Recall
                \           /
                 \         /
                  ↓       ↓
                    F1
```

### Quick Reference

| Situation                                 | Metric      |
| ----------------------------------------- | ----------- |
| Balanced classification                   | Accuracy    |
| False Positives are costly                | Precision   |
| False Negatives are costly                | Recall      |
| Both FP and FN matter                     | F1          |
| Correct Negative detection matters        | Specificity |
| Class separation across thresholds        | ROC-AUC     |
| Average regression error                  | MAE         |
| Large errors should be strongly penalized | MSE / RMSE  |
| Explained variation                       | R²          |

---

# 17. `sklearn` Functions

## Classification

```python
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    confusion_matrix,
    roc_auc_score
)
```

```python
accuracy_score(y_test, y_pred)

precision_score(y_test, y_pred)

recall_score(y_test, y_pred)

f1_score(y_test, y_pred)

confusion_matrix(y_test, y_pred)

roc_auc_score(y_test, y_prob)
```

---

## Regression

```python
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)
```

```python
mean_absolute_error(y_test, y_pred)

mean_squared_error(y_test, y_pred)

r2_score(y_test, y_pred)
```

---

# 18. Final Mental Model

```text
CLASSIFICATION
│
├── Accuracy
│   → Overall correctness
│
├── Precision
│   → Avoid False Positives
│
├── Recall
│   → Avoid False Negatives
│
├── F1
│   → Balance Precision + Recall
│
├── Specificity
│   → Identify Negatives
│
└── ROC-AUC
    → Class separation across thresholds


REGRESSION
│
├── MAE
│   → Average absolute error
│
├── MSE
│   → Penalizes large errors
│
├── RMSE
│   → Large-error penalty + original units
│
└── R²
    → Explained variation
```

# ⭐ Core Rules

$$
\text{Balanced classes + similar error costs}
\rightarrow \text{Accuracy}
$$

$$
\text{False Positive is costly}
\rightarrow \text{Precision}
$$

$$
\text{False Negative is costly}
\rightarrow \text{Recall}
$$

$$
\text{Both FP and FN matter}
\rightarrow \text{F1}
$$

$$
\text{Need class separation across thresholds}
\rightarrow \text{ROC-AUC}
$$

For regression:

```text
MAE
→ Average error

MSE
→ Penalize large errors

RMSE
→ Penalize large errors + original units

R²
→ Explained variation
```

**Core idea:** Don't choose a metric because its number is higher. Choose it based on **what the model needs to get right and which mistakes matter most**.
